timely/src/dataflow/operators/core/partition.rs

```diff
@@ -46,43 +46,41 @@ impl<G: Scope, C: Container + DrainContainer> Partition<G, C> for StreamCore<G,
+
         for _ in 0..parts {
-            let (output, stream) = builder.new_output::<CB>();
+            let (output, stream) = builder.new_output::<CB::Container>();
             outputs.push(output);
             streams.push(stream);
         }
 
+        let mut caps = BTreeMap::default();
+
         builder.build(move |_| {
-            let mut todo = vec![];
+            let mut todo: BTreeMap<_,Vec<_>> = Default::default();
             move |_frontiers| {
                 let mut handles = outputs.iter_mut().map(|o| o.activate()).collect::<Vec<_>>();
-
-                // The capability associated with each session in `sessions`.
-                let mut sessions_cap: Option<InputCapability<G::Timestamp>> = None;
-                let mut sessions = vec![];
-
+                let mut parts = BTreeMap::<u64,Vec<_>>::default();
                 while let Some((cap, data)) = input.next() {
-                    todo.push((cap, std::mem::take(data)));
+                    todo.entry(cap.time().clone()).or_default().push(std::mem::take(data));
+                    caps.insert(cap.time().clone(), cap);
                 }
-                todo.sort_unstable_by(|a, b| a.0.cmp(&b.0));
 
-                for (cap, mut data) in todo.drain(..) {
-                    if sessions_cap.as_ref().map_or(true, |s_cap| s_cap.time() != cap.time()) {
-                        sessions = handles.iter_mut().map(|h| (None, Some(h))).collect();
-                        sessions_cap = Some(cap);
+                while let Some((time, dataz)) = todo.pop_first() {
+                    let cap = caps.remove(&time).unwrap();
+                    for mut data in dataz.into_iter() {
+                        for datum in data.drain() {
+                            let (part, datum) = route(datum);
+                            parts.entry(part).or_default().push(datum);
+                        }
                     }
-                    for datum in data.drain() {
-                        let (part, datum2) = route(datum);
-
-                        let session = match sessions[part as usize] {
-                            (Some(ref mut s), _) => s,
-                            (ref mut session_slot, ref mut handle) => {
-                                let handle = handle.take().unwrap();
-                                let session = handle.session_with_builder(sessions_cap.as_ref().unwrap());
-                                session_slot.insert(session)
-                            }
-                        };
-                        session.give(datum2);
+                    while let Some((part, data)) = parts.pop_first() {
+                        for datum in data.into_iter() {
+                            c_build.push_into(datum);
+                        }
+                        while let Some(container) = c_build.finish() {
+                            handles[part as usize].give(&cap, container);
+                        }
```

Comment:
Might be better at reusing allocations if written as:
```suggestion
                        for datum in data.into_iter() {
                            c_build.push_into(datum);
                            while let Some(container) = c_build.finish() {
                                handles[part as usize].give(&cap, container);
                            }
                        }
```

Reply:
I think this would need to be `extract` rather than `finish`, and then a `finish` loop at the end? But I see what you are saying.